include/nbl/asset/filters/CFlattenRegionsStreamHashImageFilter.h

```diff
@@ -0,0 +1,220 @@
+				{
+					auto miplevel = static_cast<uint32_t>(std::distance(regions->begin(), rit));
+					auto localExtent = proxy.flatten.outImage->getMipSize(miplevel);
+					rit->bufferOffset = bufferSize;
+					rit->bufferRowLength = localExtent.x; // could round up to multiple of 8 bytes in the future
+					rit->bufferImageHeight = localExtent.y;
+					rit->imageSubresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u);
+					rit->imageSubresource.mipLevel = miplevel;
+					rit->imageSubresource.baseArrayLayer = 0u;
+					rit->imageSubresource.layerCount = parameters.arrayLayers;
+					rit->imageOffset = { 0u,0u,0u };
+					rit->imageExtent = { localExtent.x,localExtent.y,localExtent.z };
+					auto levelSize = info.roundToBlockSize(localExtent);
+					auto memsize = levelSize[0] * levelSize[1] * levelSize[2] * parameters.arrayLayers * bytesPerPixel;
+
+					assert(memsize.getNumerator() % memsize.getDenominator() == 0u);
+					bufferSize += memsize.getIntegerApprox();
+				}
+			}
+
+			auto executePerMipLevel = [&](const uint32_t miplevel)
+			{
+				/*
+					we stream-hash texels per given mip level & layer
+				*/
+
+				auto* hashers = _NBL_NEW_ARRAY(blake3_hasher, parameters.arrayLayers);
+
+				for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+					blake3_hasher_init(&hashers[layer]);
+
+				auto hash = [&hashers, &inData, &texelOrBlockByteSize](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					auto* hasher = hashers + readBlockPos.w;
+					blake3_hasher_update(hasher, inData + readBlockArrayOffset, texelOrBlockByteSize);
+				};
+
+				IImage::SSubresourceLayers subresource = { .aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u), .mipLevel = miplevel, .baseArrayLayer = 0u, .layerCount = parameters.arrayLayers }; // stick to given mip level and take all layers
+				CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { .offset = {}, .extent = { parameters.extent.width, parameters.extent.height, parameters.extent.depth } }; // cover all texels within layer range, take 0th mip level size to not clip anything at all
+				CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, parameters.format);
+
+				CBasicImageFilterCommon::executePerRegion(policy, proxy.flatten.outImage.get(), hash, regions->begin(), regions->end(), clipFunctor); // fire the hasher for layers with specified execution policy, yes you can use parallel policy here if you want at it will work
+
+				for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+				{
+					auto* out = reinterpret_cast<uint8_t*>(reinterpret_cast<CState::hash_t*>(state->scratchMemory.heap->getPointer()) + (miplevel * parameters.arrayLayers) + layer);
+					blake3_hasher_finalize(hashers + layer, out, sizeof(CState::hash_t)); // finalize hash for layer + put it to heap for given mip level
+				}
+
+				_NBL_DELETE_ARRAY(hashers, parameters.arrayLayers);
```

Comment:
you need to combine the per-layer hashes into a mip level hash here

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/711#discussion_r1677514207

Reply:
I won't change it, it can run both mips & layers with parallel policy then it would break, must finalize at the end but now using hashers from scratch memory